### PR TITLE
[FEATURE] OGR & GDAL HTTP(S) and cloud dataset support

### PR DESCRIPTION
**Pull request by nirvn**
## Description
This PR exposes - through our UI - the underlying OGR and GDAL support for HTTP(S), FTP, and cloud based datasets. 

The UI looks like this:
r
_Raster_

v
_Vector_

For the OGR provider, it was only a matter of adding a few widgets and some UI logic. For the GDAL provider, I ended up modifying the provider itself a bit to support authentication in URI.

It's really cool to see this in action. All credit goes to the GDAL folks.

Note: the only part I'm really not so sure about it how I hardcoded the protocols / drivers.  Review most welcome.

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [x] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [x] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and containt sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

### CODE REVIEW

**Comment by NathanW2** [comment]
image

File: src/providers/gdal/qgsgdalprovider.cpp [hunk 1]
```diff
@@ -3041,7 +3069,7 @@ bool QgsGdalProvider::remove()
     mGdalDataset = nullptr;
 
     CPLErrorReset();
-    CPLErr err = GDALDeleteDataset( driver, dataSourceUri().toUtf8().constData() );
+    CPLErr err = GDALDeleteDataset( driver, dataSourceUri( true ).toUtf8().constData() );
     if ( err != CPLE_None )
     {
       QgsLogger::warning( "RasterIO error: " + QString::fromUtf8( CPLGetLastErrorMsg() ) );
```
Comment (elpaso) [inline src/providers/gdal/qgsgdalprovider.cpp]: Can you check that the credentials don't get exposed in the logs in case of failure?

**Review by elpaso — APPROVED** [review]
Awesome job, thanks!

**Comment by nyalldawson** [comment]
@rouault mind looking over this?

File: src/providers/gdal/qgsgdalprovider.cpp (hunk 1 above)
Comment (nirvn) [reply src/providers/gdal/qgsgdalprovider.cpp]: @elpaso , I've made sure none of the QgsDebugMsg were calling dataSourceUri( bool ), what else should I do.

Comment (elpaso) [reply src/providers/gdal/qgsgdalprovider.cpp]: I was worried by `CPLGetLastErrorMsg()`

Comment (nirvn) [reply src/providers/gdal/qgsgdalprovider.cpp]: @elpaso , what's the alternative here, not appending CPLGetLastErrorMsg() to the warning message?

Comment (elpaso) [reply src/providers/gdal/qgsgdalprovider.cpp]: Sorry: I don't know what's in `CPLGetLastErrorMsg()`, but if there is a risk of leaking the credentials I would suggest to either try to replace them with `************` or if it's not possible to replace the native error string completely with a generic one.

**Comment by nirvn** [comment]
BTW, for those wanting to test out Amazon's AWS S3 hosted datasets, you'll have to set a few variables. It can be done easily via QGIS' Options window:
screenshot from 2018-04-23 17-03-01

File: src/providers/gdal/qgsgdalprovider.cpp (hunk 1 above)
Comment (nirvn) [reply src/providers/gdal/qgsgdalprovider.cpp]: @elpaso , I've checked a couple of CPLGetLastErrorMsg() using authcfg-based URIs, credentials aren't leaked in any of the messages I've seen.

**Comment by nirvn** [comment]
@rouault , une petite revue serait apprecie :smiley: 

I'll merge this by the weekend if I don't see any objections or change requests. Thanks.

File: src/providers/ogr/qgsogrprovider.cpp [hunk 2]
```diff
@@ -2425,6 +2425,9 @@ QString createFilters( const QString &type )
     bool kmlFound = false;
     bool dwgFound = false;
     bool dgnFound = false;
+
+    sProtocolDrivers += QLatin1String( "HTTP/HTTPS/FTP,vsicurl;AWS S3,vsis3;Google Cloud Storage,vsigs;Microsoft Azure Blob,vsiaz;Alibaba Cloud OSS,vsioss;OpenStack Swift Object Storage,vsiswift;" );
```
Comment (rouault) [inline src/providers/ogr/qgsogrprovider.cpp]: I don't think this is needed. Those are not drivers

Comment (nirvn) [reply src/providers/ogr/qgsogrprovider.cpp]: @rouault , what do you mean by "move that list closer to the UI"?

Comment (rouault) [reply src/providers/ogr/qgsogrprovider.cpp]: Your move to qgsogrsourceselect.cpp satisfies this.

**Comment by rouault** [comment]
This is a cool feature. The only thing I wonder is that it will not be obvious to people that you generally need to define some configuration options to provide the credentials. QGIS documentation will have to point to http://gdal.org/gdal_virtual_file_systems.html#gdal_virtual_file_systems_network

Another remark is that some of the VFS are only available in GDAL 2.3 (vsiaz, vsioss and vsiswift).

**Comment by nirvn** [comment]
@rouault , thanks for your review, I'll adjust the code to take into account the gdal version re azure, alibaba, and swift VFS.

I'll also add a hint in the UI regarding need for creditential option requirements.

**Comment by nirvn** [comment]
Comments addressed, will merge when travis turns green.

For the record, a useful label tip is now displayed when a cloud type is selected:
screenshot from 2018-04-26 09-57-53

**Comment by NathanW2** [comment]
Great work!

**Pull request merged**